wasm/wasm.ir/test/org/jetbrains/kotlin/wasm/ir/SpecTestRunner.kt

```diff
@@ -0,0 +1,262 @@
+            is SpecTest.Command.AssertExhaustion -> {
+            }
+            is SpecTest.Command.AssertMalformed -> {
+            }
+            is SpecTest.Command.AssertInvalid -> {
+            }
+            is SpecTest.Command.AssertUnlinkable -> {
+            }
+            is SpecTest.Command.AssertUninstantiable -> {
+            }
+            is SpecTest.Command.ActionCommand -> {
+            }
+        }
+    }
+}
+
+private fun runJsonTest(jsonFile: File, wastFile: File, wabtOptions: List<String>) {
+    require(jsonFile.isFile && jsonFile.exists())
+    val jsonText = jsonFile.readText()
+    val specTest = Json.decodeFromString(SpecTest.serializer(), jsonText)
+    val wasmDir = jsonFile.parentFile!!
+    println("Running json test ${jsonFile.path} ...")
+    runSpecTest(specTest, wasmDir, wastFile, wabtOptions)
+}
+
+val wasmTestSuitePath: String
+    get() = System.getProperty("wasm.testsuite.path")!!
+
+fun testProposal(
+    name: String,
+    wabtOptions: List<String> = listOf("--enable-all"),
+    ignoreFiles: List<String> = emptyList()
+) {
+
+    runSpecTests(name, "$wasmTestSuitePath/proposals/$name", wabtOptions, ignoreFiles)
+}
+
+
+fun runSpecTests(
+    name: String,
+    wastDirectoryPath: String,
+    wabtOptions: List<String>,
+    ignoreFiles: List<String> = emptyList()
+) {
+    // Clean and prepare output dir for spec tests
+    val specTestsDir = File("build/spec-tests/$name")
+    if (specTestsDir.exists())
+        specTestsDir.deleteRecursively()
+    specTestsDir.mkdirs()
+
+    val testSuiteDir = File(wastDirectoryPath)
+    assert(testSuiteDir.isDirectory) { "${testSuiteDir.absolutePath} is not a directory" }
+    for (file in testSuiteDir.listFiles()!!) {
+        if (file.name in ignoreFiles) {
+            println("Ignoring file: ${file.absolutePath}")
+            continue
+        }
+        if (file.isFile && file.name.endsWith(".wast")) {
+            val jsonFileName = file.withReplacedExtensionOrNull(".wast", ".json")!!.name
+            val jsonFile = File(specTestsDir, jsonFileName)
+            println("Creating JSON for ${file.path}")
+            Wabt.wast2json(file, jsonFile, *wabtOptions.toTypedArray())
+            runJsonTest(jsonFile, file, wabtOptions)
+        }
+    }
+}
+
+
+fun testWasmFile(wasmFile: File, dirName: String) {
+    val testName = wasmFile.nameWithoutExtension
+
+    fun newFile(suffix: String): File =
+        File("build/spec-tests/tmp/$dirName/${testName}_$suffix")
+            .also {
+                it.parentFile.mkdirs()
+                it.createNewFile()
+            }
+
+    println("Testing wasm file : ${wasmFile.absolutePath} ... ")
+    val module = fileToWasmModule(wasmFile)
+    module.calculateIds()
+    val kotlinTextFormat = module.toTextFormat()
+    val kotlinBinaryFormat = module.toBinaryFormat()
+
+    val kotlinTextFile = newFile("kwt.wat")
+    kotlinTextFile.writeText(kotlinTextFormat)
+    val kotlinBinaryFile = newFile("kwt.wasm")
```

Comment:
Wouldn't work. There are various ways to encode the same Wasm module, for example it can encode the same number using LEB128 with different number of bytes.

Reply:
isn't there only one way for the shortest in LEB128?

Reply:
Probably another case are some floating point constants, is there anything else?

Reply:
There is a sing shortest LEB128 for each number, yes, but it doesn't have to be shortest. And Wasm linkers usually use max-windth encoding of a placeholder for late binding.

There are various other ways where binary of the same mode wouldnt match, like omitting/including empty section, encoding optimization for multiple anjacent local variables of the same type, etc.